compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,677 @@
+//! This module provides a framework on top of the normal MIR dataflow framework to simplify the
+//! implementation of analyses that track the values stored in places of interest.
+//!
+//! The default methods of [`ValueAnalysis`] (prefixed with `super_` instead of `handle_`)
+//! provide some behavior that should be valid for all abstract domains that are based only on the
+//! value stored in a certain place. On top of these default rules, an implementation should
+//! override some of the `handle_` methods. For an example, see `ConstAnalysis`.
+//!
+//! An implementation must also provide a [`Map`]. Before the anaylsis begins, all places that
+//! should be tracked during the analysis must be registered. The set of tracked places cannot be
+//! changed during the analysis.
```

Comment:
I changed the wording a little and also documented how references are tracked. Let me know if you think that this is still unclear.

Reply:
The added comment is indeed helpful, but does not really address my main point. What I mean is that the `Place` struct (instances of which you are - very roughly speaking - tracking), is really a place *expression*. In other words, it is a set of instructions for how to compute a place. This is distinct from the *result* of computing that place (a "place value") which is essentially some bytes in memory. For a simple example, consider
```rust
let _ = *p;
p = whatever();
let _ = *p;
```
The first and third line both contain the same place expression (`*p`) but that place expression refers to a different location in memory in each of those lines. This is the thing that I would like to see acknowledged.

To be clear, as far as I can tell all your code handles this correctly, since you flood all children of `p` on an assignment to `p`. This is just a fairly significant caveat in the soundness proof, that's why I bring it up.

Reply:
Ah, now I see what you mean. Yes, thats a good point. I added some comments regarding the terminology. In order to keep the meaning of `mir::Place`, I would use "place" to refer to "place expression" and "object" to refer to the underlying object for now. But I'm open to suggestions for better terms.